applications/AdjointFluidApplication/custom_schemes/adjoint_steady_scheme.h

```diff
@@ -206,10 +225,27 @@ class AdjointSteadyScheme : public Scheme<TSparseSpace, TDenseSpace>
     {
         KRATOS_TRY
 
-        for (auto it = rDofSet.begin(); it != rDofSet.end(); ++it)
-            if (it->IsFree() == true)
-                it->GetSolutionStepValue() +=
-                    TSparseSpace::GetValue(rDx, it->EquationId());
+        Communicator& rComm = rModelPart.GetCommunicator();
+
+        if (rComm.TotalProcesses() == 1)
+        {
+            for (auto it = rDofSet.begin(); it != rDofSet.end(); ++it)
+                if (it->IsFree() == true)
+                    it->GetSolutionStepValue() +=
+                        TSparseSpace::GetValue(rDx, it->EquationId());
+        }
+        else
+        {
+            for (auto it = rDofSet.begin(); it != rDofSet.end(); ++it)
+                if (it->GetSolutionStepValue(PARTITION_INDEX) == rComm.MyPID())
+                    if (it->IsFree() == true)
+                        it->GetSolutionStepValue() +=
+                            TSparseSpace::GetValue(rDx, it->EquationId());
+
+            // todo: add a function Communicator::SynchronizeDofVariables() to
+            // reduce communication here.
+            rComm.SynchronizeNodalSolutionStepsData();
```

Comment:
I don't really understand why the original version was deleted... If I understand this right rDx is a distributed vector that, at this point, is already synchronized, and you should be able to access any given row from all processes that know that DOF. You should be able to use your "serial" code everywhere (no need to check for PARTITION_INDEX) and, since all processes access all their nodes, this would allow you to avoid the Synchronize